src/Service.Host/client/src/components/salesPackages/SalesPackage.js

```diff
@@ -0,0 +1,275 @@
+} from '@linn-it/linn-form-components-library';
+import Page from '../../containers/Page';
+
+function SalesPackage({
+    item,
+    editStatus,
+    setEditStatus,
+    errorMessage,
+    itemId,
+    update,
+    history,
+    add,
+    loading,
+    snackbarVisible,
+    setSnackbarVisible
+}) {
+    const [salesPackage, setSalesPackage] = useState({});
+    const [prevSalesPackage, setPrevSalesPackage] = useState({});
+    const [showAddRow, setShowAddRow] = useState(false);
+    const [newElement, setNewElement] = useState({});
+
+    const creating = () => editStatus === 'create';
+    const editing = () => editStatus === 'edit';
+    const viewing = () => editStatus === 'view';
+    const cursor = {
+        cursor: 'pointer'
+    };
+    useEffect(() => {
+        if (!creating() && item !== prevSalesPackage) {
+            setSalesPackage(item);
+            setPrevSalesPackage(item);
+        }
+    });
+
+    const showFieldsToAddElement = () => {
+        setShowAddRow(true);
+    };
+
+    const inputInvalid = () => !salesPackage.salesPackageId;
+
+    const handleSaveClick = () => {
+        if (editing()) {
+            if (newElement.elementType) {
+                const { elements } = salesPackage;
+                elements.push(newElement);
+                setSalesPackage({ ...salesPackage, elements });
+            }
+            update(itemId, salesPackage);
+            setEditStatus('view');
+            setNewElement({});
+            setShowAddRow(false);
+        } else if (creating()) {
+            add(salesPackage);
+        }
+    };
+
+    const handleCancelClick = () => {
+        setSalesPackage(item);
+        setEditStatus('view');
+    };
+
+    const handleBackClick = () => {
+        history.push('/products/maint/sales-packages');
+    };
+
+    const handleFieldChange = (propertyName, newValue) => {
+        setEditStatus('edit');
+        setSalesPackage({ ...salesPackage, [propertyName]: newValue });
+    };
+
+    const handleNewElement = (propertyName, newValue) => {
+        setEditStatus('edit');
+        setNewElement({ ...newElement, [propertyName]: newValue });
+    };
+
+    const handleElementChange = (propertyName, newValue) => {
+        setEditStatus('edit');
+        const splitIndex = propertyName.indexOf(',');
+        const index = propertyName.slice(0, splitIndex);
+        const prop = propertyName.slice(splitIndex + 1);
+        const { elements } = salesPackage;
+        elements[index][prop] = newValue;
+        setSalesPackage({ ...salesPackage, elements });
+    };
+
+    return (
+        <Page>
+            <Grid container spacing={24}>
+                <Grid item xs={12}>
+                    {creating() ? (
+                        <Title text="Add Sales Package" />
+                    ) : (
+                        <Title text="Sales Package Details" />
+                    )}
+                </Grid>
+                {loading || (!item && !creating()) ? (
```

Comment:
I would change your loading logic to check whether the salesPackage has been loaded into state, something like: 

`                {loading || (!salesPackage && !creating()) ? ( ` ...render loading


It can be the case that the component tries to render before the item is in state... Without a check like this you could end up with 'cannot read property' of null style errors

Reply:
👍